### PR TITLE
Adds a new command to randomize (shuffle) tracks

### PR DESCRIPTION
**Pull request by jarv**
Command added to the library, playlist and queue view for
feature request https://github.com/cmus/cmus/issues/231.

### CODE REVIEW

**Comment by flyingmutant** [comment]
Thanks for the PR! I'll add some comments inline.

File: command_mode.c [hunk 1]
```diff
@@ -1352,6 +1352,26 @@ static void cmd_pwd(char *arg)
...
+		editable_rand(&pl_editable);
+		break;
+	case QUEUE_VIEW:
+		editable_rand(&pq_editable);
+		break;
+	}
+	editable_unlock();
+}
+
+
```
Comment (flyingmutant) [inline command_mode.c]: Please keep whitespace between functions consistent with status quo in files you change, which is 1 empty line, not 2.

File: track.c [hunk 2]
```diff
@@ -250,6 +250,57 @@ void sorted_list_remove_track(struct list_head *head, struct rb_root *tree_root,
 	list_del(&track->node);
 }
 
+static void shuffle_array(void *array, size_t n, size_t size)
```
Comment (flyingmutant) [inline track.c]: I think it is better to move this function to `misc.c`.

File: track.c [hunk 3]
```diff
@@ -250,6 +250,57 @@ void sorted_list_remove_track(struct list_head *head, struct rb_root *tree_root,
 	list_del(&track->node);
 }
 
+static void shuffle_array(void *array, size_t n, size_t size) 
+{
+	char tmp[size];
```
Comment (flyingmutant) [inline track.c]: Hm, is this the first place in cmus where var-sized arrays are used? I am not 100% sure I am comfortable with it for some reason.

File: track.c [hunk 4]
```diff
@@ -250,6 +250,57 @@ void sorted_list_remove_track(struct list_head *head, struct rb_root *tree_root,
...
+	char *arr = array;
+	size_t stride = size * sizeof(char);
```
Comment (flyingmutant) [inline track.c]: `sizeof(char)` is 1.

File: track.c [hunk 5]
```diff
@@ -250,6 +250,57 @@ void sorted_list_remove_track(struct list_head *head, struct rb_root *tree_root,
...
+	if (n > 1) {
+		size_t i;
+		for (i = 0; i < n - 1; ++i) {
```
Comment (flyingmutant) [inline track.c]: These 3 lines can be replaced by something like `for (ssize_t i = 0; i < (ssize_t)n - 1; ++i)`.

File: track.c [hunk 6]
```diff
@@ -250,6 +250,57 @@ void sorted_list_remove_track(struct list_head *head, struct rb_root *tree_root,
...
+		}
+	}
+}
+
+
+void rand_list_rebuild(struct list_head *head, struct rb_root *tree_root)
+{
+	struct list_head *item, *tmp;
+	struct rb_root tmp_tree = RB_ROOT;
+	struct simple_track **track_list;
```
Comment (flyingmutant) [inline track.c]: `track_list` is really an array, not a list.

File: track.c [hunk 7]
```diff
@@ -250,6 +250,57 @@ void sorted_list_remove_track(struct list_head *head, struct rb_root *tree_root,
...
+	static const sort_key_t empty_sort_keys[] = { SORT_INVALID };
+
+	unsigned int len = 0, track_cnt = 0, i =0;
+
+	list_for_each(item, head) {
+		len++;
+	}
+	track_list = xmalloc(len * sizeof(struct simple_track *));
```
Comment (flyingmutant) [inline track.c]: `sizeof(track_list[0])` is much better.

File: track.c [hunk 8]
```diff
@@ -250,6 +250,57 @@ void sorted_list_remove_track(struct list_head *head, struct rb_root *tree_root,
...
+
+	LIST_HEAD(tmp_head);
+	list_for_each_safe(item, tmp, head) {
+		struct simple_track *track = to_simple_track(item);
+		sorted_list_remove_track(head, tree_root, track);
+		track_list[track_cnt] = track;
+		track_cnt++;
+	}
+	shuffle_array(track_list, len, sizeof(struct simple_track *));
```
Comment (flyingmutant) [inline track.c]: Same here – no need to hardcode exact types in `sizeof`.

File: track.c [hunk 9]
```diff
@@ -250,6 +250,57 @@ void sorted_list_remove_track(struct list_head *head, struct rb_root *tree_root,
...
+	for (i=0; i<len; i++) {
```
Comment (flyingmutant) [inline track.c]: No need to declare `i` beforehand – just do `for (int i = 0; …)`.

File: track.c [hunk 10]
```diff
@@ -250,6 +250,57 @@ void sorted_list_remove_track(struct list_head *head, struct rb_root *tree_root,
...
+		sorted_list_add_track(&tmp_head, &tmp_tree, track_list[i], empty_sort_keys, 0);
+	}
+	free(track_list);
+
+	tree_root->rb_node = tmp_tree.rb_node;
+	_list_add(head, tmp_head.prev, tmp_head.next);
+
```
Comment (flyingmutant) [inline track.c]: Extra newline.

**Comment by jarv** [comment]
I've gone ahead and addressed your comments in a new commit (let me know if you would like me to squash).  For the var-sized array I was hesitant to allocate MAX_TRACKS and avoided doing something more complex for the shuffle. Let me know what you think if you are still open to taking this. I use CMUs all the time and randomizing tracks has been very useful for me when DJing from a playlist.

**Comment by jarv** [comment]
Updated to fix the last comments

**Comment by flyingmutant** [comment]
Rebased and merged in a10327e7dbb07fcc68e3fd08ff519f8bffb2353b, thanks a lot!

**Pull request closed without merging**